File: Data-Structure-I/5-Tree/P102.cpp
// Date: Sun, Nov 21, 2021
// 102. Binary Tree Level Order Traversal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	// my own solution
	// time complexity: O(n) where n is the number of nodes
	// space complexity: O(n)
    vector<vector<int>> levelOrder(TreeNode* root) {
    	if (!root)
    		return {};

        vector<vector<int>> ans;

        queue<TreeNode*> my_queue;
        my_queue.push(root);

        while (!my_queue.empty()) {
        	queue<TreeNode*> tmp_queue;
        	vector<int> tmp_vector;
        	while(!my_queue.empty()) {
        		TreeNode* current = my_queue.front();
        		my_queue.pop();
        		if (current->left)
        			tmp_queue.push(current->left);
        		if (current->right)
        			tmp_queue.push(current->right);
				tmp_vector.push_back(current->val);
        	}
        	my_queue = tmp_queue;
        	ans.push_back(tmp_vector);
        }
        return ans;
    }
};

File: LeetCoding-Challenge/July-2022/P102.java
package leetcodingchallenge.july2022;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Date: July 13, 2022 - Already solve this problem in C++ on Nov 21, 2021
 * 102. Binary Tree Level Order Traversal
 * https://leetcode.com/problems/binary-tree-level-order-traversal/
 */

public class P102 {
    // time & space: O(n)
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int n = queue.size();
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                TreeNode cur = queue.pollFirst();
                list.add(cur.val);
                if (cur.left != null) queue.add(cur.left);
                if (cur.right != null) queue.add(cur.right);
            }
            ans.add(list);
        }
        return ans;
    }

    // Definition for a binary tree node.
    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
